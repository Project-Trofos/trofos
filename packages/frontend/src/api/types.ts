export type Project = {
  id: number;
  pname: string;
  pkey: string | null;
  description: string | null;
  course_id: number | null;
  public: boolean;
  created_at: string;
  course?: Course;
  is_archive: boolean | null;
};

export type Course = {
  id: number;
  code: string;
  startYear: number;
  startSem: number;
  endYear: number;
  endSem: number;
  cname: string;
  description: string | null;
  public: boolean;
  created_at: string;
  shadow_course: boolean;
  is_archive: boolean | null;
};

export type User = {
  user_email: string;
  user_display_name: string;
  user_id: number;
  projects: Project[];
  courses: CourseRoles[];
  basicRoles: BasicRoles[];
};

export type BasicRoles = {
  user_email: string;
  role_id: number;
};

export type CourseRoles = {
  id: number;
  user_email: string;
  role_id: number;
  course_id: number;
};

export type CreateUserRequest = {
  userEmail: string;
  newPassword: string;
};

export type RegisterUser = {
  userEmail: string;
  newPassword: string;
  userDisplayName: string;
};

export type UserCourseRoleRequest = {
  id: number;
  userRole: number;
  userId: number;
};

export type Role = {
  role_name: string;
  id: number;
};

export type Action = {
  action: string;
};

export type ActionOnRole = {
  id: number;
  action: string;
};

export type ActionsOnRoles = {
  id: number;
  role_name: string;
  actions: Action[];
};

export type ProjectData = Project & {
  course: Course | null;
  users: UserData[];
  sprints: {
    id: number;
    name: string;
  }[];
  backlogStatuses: Omit<BacklogStatusData, 'project_id'>[];
  telegramChannelLink: string;
  owner_id: number | null;
};

export type UserData = {
  user: Pick<User, 'user_id' | 'user_email' | 'user_display_name' | 'courses'>;
};

export type ScrumBoardUserData = {
  user: {
    user_id: number | null;
    user_display_name: string;
    user_email: string;
  };
};

export type Milestone = {
  id: number;
  name: string;
  start_date: string;
  deadline: string;
  course_id: number;
  created_at: string;
};

export type Announcement = {
  id: number;
  title: string;
  content: string;
  user_id: number;
  created_at: string;
  updated_at?: string;
};

// CourseData type returned by BE api. We perform transformResponse to convert the data into type CourseData
export type CourseDataResponse = Omit<CourseData, 'users'> & {
  courseRoles: UserData[];
};

export type CourseData = Course & {
  milestones: Milestone[];
  announcements: Announcement[];
  users: UserData[];
};

export type UserOnRolesOnCourse = {
  id: number;
  user_id: number;
  role_id: number;
  course_id: number;
  role: {
    id: number;
    role_name: string;
  };
};

export type BacklogStatusData = {
  name: string;
  type: 'todo' | 'in_progress' | 'done';
  project_id: number;
  order: number;
};

export type Comment = {
  comment_id: number;
  backlog_id: number;
  project_id: number;
  commenter_id: number;
  content: string;
  created_at: string;
  updated_at: string | null;
  commenter: {
    created_at: string;
    project_id: string;
    user: {
      user_id: number;
      user_email: string;
      user_display_name: string;
    };
  };
};

export type CommentFieldsType = {
  projectId: number;
  backlogId: number;
  commenterId: number;
  content: string;
};

export enum BacklogStatus {
  TODO = 'To do',
  DONE = 'Done',
}

export type BacklogPriority = 'very_high' | 'high' | 'medium' | 'low' | 'very_low' | null;

export type Backlog = {
  backlog_id: number;
  summary: string;
  type: 'story' | 'task' | 'bug';
  priority: BacklogPriority;
  reporter_id: number;
  assignee_id: number | null;
  sprint_id: number | null;
  points: number | null;
  description: string | null;
  project_id: number;
  status: BacklogStatus | string;
  assignee: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
  epic?: Epic | null;
};

export type Epic = {
  epic_id: number;
  project_id: number;
  name: string;
  description: string | null;
};

export type DefaultBacklog = {
  backlog_id?: number;
  summary?: string;
  type?: 'story' | 'task' | 'bug';
  priority?: BacklogPriority;
  reporter_id?: number;
  assignee_id?: number | null;
  sprint_id?: number | null;
  points?: number | null;
  description?: string | null;
  project_id?: number;
  status?: BacklogStatus | string;
  assignee?: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
};

export type BacklogUpdatePayload = {
  projectId: number;
  backlogId: number;
  srcSprintId?: number | null;
  fieldToUpdate: Partial<Backlog>;
};

export enum BacklogHistoryType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
}

export type BacklogHistory = {
  history_type: BacklogHistoryType;
  date: string;
} & Omit<Backlog, 'assignee' | 'summary' | 'description'>;

export type UpdateUserRolePayload = {
  userId: number;
  newRoleId: number;
};

export type ProjectGitLink = {
  project_id: number;
  repo: string;
};

export type ProjectGitLinkData = {
  projectId: number;
  repoLink: string;
};

export type ProjectUserSettings = {
  project_id: number;
  user_id: number;
  email_notification: boolean;
};

export type Settings = {
  current_year: number;
  current_sem: number;
};

export type RetrospectiveVote = {
  id: number;
  retro_id: number;
  user_id: number;
  type: RetrospectiveVoteType;
};

export type Retrospective = {
  id: number;
  sprint_id: number;
  content: string;
  type: RetrospectiveType;
  score: number;
  votes: RetrospectiveVote[];
  is_action_taken: boolean;
};

export enum RetrospectiveType {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  ACTION = 'action',
}

export enum RetrospectiveVoteType {
  UP = 'up',
  DOWN = 'down',
}

export type CourseImportCsvPayload = {
  courseId: number;
  payload: FormData;
};

export type Feedback = {
  id: number;
  user_id?: number;
  user: {
    user_email: string;
    user_display_name: string;
  };
  sprint_id: number;
  content: string;
  created_at: string;
  updated_at?: string;
};

export type OAuth2Payload = {
  code: string;
  state: string;
  callbackUrl: string;
};

export type Invite = {
  project_id: number;
  email: string;
  expiry_date: Date;
};

export type UserApiKey = {
  id: number;
  user_id: number;
  api_key: string;
  created_at: string;
  last_used: string | null;
  active: boolean;
};

export type UserGuideQueryResponse = {
  answer: string;
  links: Array<string>;
};

export type FeatureFlag = {
  feature_name: string;
  active: boolean;
};

export type ProjectAssignment = {
  id: number;
  targetProjectId: number;
  sourceProjectId: number;
};
