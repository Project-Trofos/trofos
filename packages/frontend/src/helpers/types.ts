export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
