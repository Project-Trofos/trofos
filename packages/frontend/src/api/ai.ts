import trofosApiSlice from '.';
import { UserGuideQueryResponse, UserGuideRecommendation } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    answerUserGuideQuery: builder.mutation<UserGuideQueryResponse, { query: string; isEnableMemory: boolean }>({
      query: ({ query, isEnableMemory }) => ({
        url: `ai/userGuideQuery`,
        credentials: 'include',
        method: 'POST',
        body: { query, isEnableMemory },
      }),
    }),
    recommendUserGuideSections: builder.mutation<UserGuideRecommendation[], void>({
      query: () => ({
        url: `ai/recommendUserGuide`,
        method: 'POST',
        credentials: 'include',
      }),
    }),
  }),
});

export const { useAnswerUserGuideQueryMutation, useRecommendUserGuideSectionsMutation } = extendedApi;
