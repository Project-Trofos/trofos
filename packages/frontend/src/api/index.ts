import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
