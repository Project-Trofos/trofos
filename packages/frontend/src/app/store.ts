import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
