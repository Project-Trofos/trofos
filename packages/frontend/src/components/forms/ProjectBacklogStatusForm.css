.project-backlog-status-draggable {
  padding-right: 22px;
  background-color: lightblue;
  display: flex;
  margin-bottom: 16px;
  position: relative;
}

.project-backlog-status-draggable .ant-form-item {
  margin-bottom: 0;
  width: 100%;
}

.project-backlog-status-delete-button {
  display: flex;
  align-items: center;
  padding: 0 2px;
  cursor: pointer;
}

.project-backlog-status-delete-button span {
  visibility: hidden;
}

.project-backlog-status-delete-button:hover {
  background-color: red;
}

.project-backlog-status-delete-button:hover span {
  visibility: visible;
}
