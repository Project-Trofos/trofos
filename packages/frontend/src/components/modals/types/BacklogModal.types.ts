export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
