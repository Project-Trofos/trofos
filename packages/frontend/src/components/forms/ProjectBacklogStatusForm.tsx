import React, { useEffect, useState } from 'react';
import { Button, Form, Input, message } from 'antd';
import { CloseOutlined } from '@ant-design/icons';
import { useParams } from 'react-router-dom';
import { DragDropContext, Draggable } from 'react-beautiful-dnd';
import StrictModeDroppable from '../dnd/StrictModeDroppable';
import { BacklogStatusData } from '../../api/types';
import {
  useCreateBacklogStatusMutation,
  useDeleteBacklogStatusMutation,
  useUpdateBacklogStatusMutation,
  useUpdateBacklogStatusOrderMutation,
} from '../../api/project';
import './ProjectBacklogStatusForm.css';
import { sortBacklogStatus } from '../../helpers/sortBacklogStatus';

type StatuesesPropType = Omit<BacklogStatusData, 'projectId'>[];

export default function ProjectBacklogStatusForm(props: { statuses: StatuesesPropType }) {
  const { statuses } = props;

  const [form] = Form.useForm();

  const params = useParams();
  const projectId = Number(params.projectId);

  const [createBacklogStatus] = useCreateBacklogStatusMutation();
  const [updateBacklogStatus] = useUpdateBacklogStatusMutation();
  const [updateBacklogStatusOrder] = useUpdateBacklogStatusOrderMutation();
  const [deleteBacklogStatus] = useDeleteBacklogStatusMutation();

  const [isAddingStatus, setIsAddingStatus] = useState(false);

  const handleBacklogStatusUpdate = async (e: React.FocusEvent<HTMLInputElement, Element>) => {
    if (!e?.target?.id || !e?.target?.value || e.target.id === e.target.value) {
      return;
    }

    const backlogStatusToUpdate = {
      projectId,
      currentName: e.target.id,
      updatedName: e.target.value,
    };

    try {
      await updateBacklogStatus(backlogStatusToUpdate).unwrap();
      message.success({ content: 'Backlog status updated', key: 'backlogStatusUpdateMessage' });
    } catch (err) {
      message.error({ content: 'Failed to update backlog status', key: 'backlogStatusUpdateMessage' });
      console.error(err);
    }
  };

  const reorder = (list: StatuesesPropType, startIndex: number, endIndex: number): StatuesesPropType => {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);

    return result.map((status, index) => {
      // eslint-disable-next-line no-param-reassign
      status.order = index + 1;
      return status;
    });
  };

  const [statusOrder, setStatusOrder] = useState<StatuesesPropType>([]);
  const [inProgressOrder, setInProgressOrder] = useState<StatuesesPropType>([]);

  useEffect(() => {
    const sortedStatuses = sortBacklogStatus(statuses);
    setStatusOrder(sortedStatuses);
    setInProgressOrder(sortedStatuses.filter((status) => status.type === 'in_progress'));
  }, [statuses]);

  const onDragEnd = async (result: any) => {
    if (!result.destination) {
      return;
    }

    if (result.destination.index === result.source.index) {
      return;
    }

    const updatedStatusOrder: StatuesesPropType = reorder(
      inProgressOrder,
      result.source.index,
      result.destination.index,
    );

    const payload = {
      projectId,
      updatedStatuses: updatedStatusOrder,
    };

    try {
      await updateBacklogStatusOrder(payload).unwrap();
      setInProgressOrder(updatedStatusOrder);
      message.success({ content: 'Status order updated', key: 'backlogStatusOrderUpdateMessage' });
    } catch (err) {
      message.error({ content: 'Failed to update status order', key: 'backlogStatusOrderUpdateMessage' });
      console.error(err);
    }
  };

  const handleAddStatus = async (e: React.FocusEvent<HTMLInputElement, Element>) => {
    if (!e?.target?.value) {
      setIsAddingStatus(false);
      return;
    }

    const statusToCreate = {
      projectId,
      name: e.target.value,
    };

    try {
      await createBacklogStatus(statusToCreate).unwrap();
      setIsAddingStatus(false);
      message.success({ content: 'Backlog status added', key: 'addBacklogStatusMessage' });
    } catch (err) {
      message.error({ content: 'Failed to add backlog status', key: 'addBacklogStatusMessage' });
      console.error(err);
    }
  };

  const handleDeleteStatus = async (name: string) => {
    if (!name) {
      return;
    }

    const statusToDelete = {
      projectId,
      name,
    };

    try {
      await deleteBacklogStatus(statusToDelete).unwrap();
      message.success({ content: 'Backlog status deleted', key: 'deleteBacklogStatusMessage' });
    } catch (err) {
      message.error({
        content: 'Failed to delete backlog status. Please check if there are any backlogs assigned with this status.',
        key: 'deleteBacklogStatusMessage',
      });
      console.error(err);
    }
  };

  return (
    <>
      <Form form={form}>
        <Form.Item
          name={statusOrder?.[0]?.name}
          initialValue={statusOrder?.[0]?.name}
          key={statusOrder?.[0]?.name}
          rules={[{ required: true }]}
        >
          <Input onBlur={handleBacklogStatusUpdate} />
        </Form.Item>
        <DragDropContext onDragEnd={onDragEnd}>
          <StrictModeDroppable droppableId="backlogStatusList">
            {(provided) => (
              <div ref={provided.innerRef} {...provided.droppableProps}>
                {inProgressOrder.map((status, index) => (
                  <Draggable key={status.name} draggableId={status.name} index={index}>
                    {(draggableProvided) => (
                      <div
                        className="project-backlog-status-draggable"
                        ref={draggableProvided.innerRef}
                        {...draggableProvided.draggableProps}
                        {...draggableProvided.dragHandleProps}
                      >
                        <Form.Item
                          name={status.name}
                          initialValue={status.name}
                          key={status.name}
                          rules={[{ required: true }]}
                        >
                          <Input onBlur={handleBacklogStatusUpdate} />
                        </Form.Item>
                        {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
                        <div
                          className="project-backlog-status-delete-button"
                          onClick={() => handleDeleteStatus(status.name)}
                        >
                          <CloseOutlined />
                        </div>
                      </div>
                    )}
                  </Draggable>
                ))}
                {provided.placeholder}
              </div>
            )}
          </StrictModeDroppable>
        </DragDropContext>
        {isAddingStatus && (
          <Form.Item rules={[{ required: true }]}>
            <Input onBlur={handleAddStatus} placeholder="Enter status name..." />
          </Form.Item>
        )}
        <Form.Item
          name={statusOrder?.[statusOrder.length - 1]?.name}
          initialValue={statusOrder?.[statusOrder.length - 1]?.name}
          key={statusOrder?.[statusOrder.length - 1]?.name}
          rules={[{ required: true }]}
        >
          <Input onBlur={handleBacklogStatusUpdate} />
        </Form.Item>
      </Form>
      <Button type="primary" onClick={() => setIsAddingStatus(true)}>
        Add
      </Button>
    </>
  );
}
