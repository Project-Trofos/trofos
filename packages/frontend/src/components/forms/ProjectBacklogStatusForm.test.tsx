import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BacklogStatusData } from '../../api/types';
import store from '../../app/store';
import ProjectBacklogStatusForm from './ProjectBacklogStatusForm';

describe('ProjectBacklogStatusForm test', () => {
  const mockStatuses: BacklogStatusData[] = [
    {
      project_id: 123,
      name: 'To do',
      type: 'todo',
      order: 1,
    },
    {
      project_id: 123,
      name: 'In progress',
      type: 'in_progress',
      order: 1,
    },
    {
      project_id: 123,
      name: 'Done',
      type: 'done',
      order: 1,
    },
  ];

  function renderForm() {
    return render(
      <Provider store={store}>
        <ProjectBacklogStatusForm statuses={mockStatuses} />
      </Provider>,
    );
  }

  describe('when rendering', () => {
    it('should render according to the snapshot', () => {
      const { baseElement } = renderForm();
      expect(baseElement).toMatchSnapshot();
    });
  });
});
