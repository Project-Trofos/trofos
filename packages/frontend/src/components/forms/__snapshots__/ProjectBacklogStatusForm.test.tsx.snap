// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectBacklogStatusForm test > when rendering > should render according to the snapshot 1`] = `
<body>
  <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input ant-input-outlined"
                  id="To do"
                  type="text"
                  value="To do"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="backlogStatusList"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          class="project-backlog-status-draggable"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="In progress"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="In progress"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <div
            class="ant-form-item"
          >
            <div
              class="ant-row ant-form-item-row"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <span
                      class="ant-input-group-wrapper ant-input-group-wrapper-outlined"
                    >
                      <span
                        class="ant-input-wrapper ant-input-group"
                      >
                        <span
                          class="ant-input-group-addon"
                        >
                          <div
                            class="project-backlog-status-move-button"
                          >
                            <span
                              aria-label="drag"
                              class="anticon anticon-drag"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="drag"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M909.3 506.3L781.7 405.6a7.23 7.23 0 00-11.7 5.7V476H548V254h64.8c6 0 9.4-7 5.7-11.7L517.7 114.7a7.14 7.14 0 00-11.3 0L405.6 242.3a7.23 7.23 0 005.7 11.7H476v222H254v-64.8c0-6-7-9.4-11.7-5.7L114.7 506.3a7.14 7.14 0 000 11.3l127.5 100.8c4.7 3.7 11.7.4 11.7-5.7V548h222v222h-64.8c-6 0-9.4 7-5.7 11.7l100.8 127.5c2.9 3.7 8.5 3.7 11.3 0l100.8-127.5c3.7-4.7.4-11.7-5.7-11.7H548V548h222v64.8c0 6 7 9.4 11.7 5.7l127.5-100.8a7.3 7.3 0 00.1-11.4z"
                                />
                              </svg>
                            </span>
                          </div>
                        </span>
                        <input
                          aria-required="true"
                          class="ant-input ant-input-outlined"
                          id="In progress"
                          type="text"
                          value="In progress"
                        />
                        <span
                          class="ant-input-group-addon"
                        >
                          <div
                            class="project-backlog-status-delete-button"
                          >
                            <span
                              aria-label="close"
                              class="anticon anticon-close"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close"
                                fill="currentColor"
                                fill-rule="evenodd"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                                />
                              </svg>
                            </span>
                          </div>
                        </span>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input ant-input-outlined"
                  id="Done"
                  type="text"
                  value="Done"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
    <button
      class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
      type="button"
    >
      <span>
        Add
      </span>
    </button>
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="rbd-announcement-0"
    style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
  />
  <div
    id="rbd-hidden-text-0-hidden-text-0"
    style="display: none;"
  >
    
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

  </div>
</body>
`;
